### ISSUE TITLE
Excessive memory usage on PHP7

### ISSUE THREAD

**Issue by vojta** [wcgallego/pecl-gearman#7]
I get "Allowed memory size of 134217728 bytes exhausted (tried to allocate 3472891254821631824 bytes) " 
This doesn't seem OK to me :)

How to reproduce:

gearman_worker.php:

```
<?php

error_reporting(E_ALL | E_STRICT);

$fRun = function ($task) {

    $taskData = $task->workload();

    try {
        $data = $task->workload();
        print 'Task OK ' . $task->unique() . ' len: ' . strlen($data) . PHP_EOL;
        $task->sendComplete(strrev($data));

    } catch (\Exception $e) {
            $str = $e->__toString();
            $task->sendException($str);

            return \GEARMAN_ERRNO;
        }

    return GEARMAN_SUCCESS;
};

print "Worker started, pid " . getmypid() . PHP_EOL;

$worker = new GearmanWorker();
$worker->addServer('127.0.0.1', 4730);

$worker->addFunction('myDefaultMethod', $fRun);

$worker->work();
if ($worker->returnCode() != GEARMAN_SUCCESS) {
    print("return code: " . $worker->returnCode() . PHP_EOL);
} else {
    print 'Worker OK' . PHP_EOL;
}
```

gearman_client.php:

```
<?php

error_reporting(E_ALL | E_STRICT);

class Client
{
    /**
     * \GearmanClient
     */
    protected $gc;

    public function __construct()
    {
        $gc = new \GearmanClient();
        $gc->addServer();
        $gc->setTimeout(1000);

        $this->gc = $gc;
        print 'client created' . PHP_EOL;
    }

    public function ping($str)
    {
        return $this->gc->ping($str);
    }

    public function runJobs()
    {
        print 'runJobs called' . PHP_EOL;

        var_dump($this->gc->setCompleteCallback([$this, 'jobCallback']));
        var_dump($this->gc->setExceptionCallback([$this, 'exceptionCallback']));
        var_dump($this->gc->setFailCallback([$this, 'failCallback']));

        for ($i = 0; $i < 100; $i++) {
            print('calling addTask()' . PHP_EOL);
            $jobStr = str_repeat((string) rand(), 1000);
            $jobLen = strlen($jobStr);
            print 'job length: ' . $jobLen . PHP_EOL;

            for ($j = 0; $j < 5; $j++) {
                $gTask = $this->gc->addTask('myDefaultMethod', $j . '_' . $jobStr, null, 'uniq' . $i . '_' .  $j);
            }

            if ($gTask instanceof \GearmanTask) {
                print 'OK' . PHP_EOL;
            }

            print 'runTasks call' . PHP_EOL;
            $this->gc->runTasks();
        }
    }

    public function jobCallback($task)
    {
        print 'jobCallback called' . PHP_EOL;
        var_dump(strlen($task->data()));

        return GEARMAN_SUCCESS;
    }

    public function exceptionCallback($task)
    {
        print 'exceptionCallback()' . PHP_EOL;
        var_dump($task->data());
        throw new \Exception($task->data());
    }

    public function failCallback($task)
    {
        print 'failCallback()' . PHP_EOL;
        var_dump($task->data());
    }
}



// ============= RUN =============

$bgc = new \Client();

$pingResult = $bgc->ping('xyz');
print 'ping called with result: ' . var_export($pingResult, 1) . PHP_EOL;

$bgc->runJobs();
```

To reproduce, just start the workers in parallel:

```
yes | xargs -n1 -P4 -I {} bash -c "php70 ./gearman_worker.php || echo 'WORKER FAILED'"
```

and run the client:

```
$ php70 -d memory_limit=128M ./gearman_client.php; echo $?
client created
ping called with result: true
runJobs called
bool(true)
bool(true)
bool(true)
calling addTask()
job length: 9000
OK
runTasks call
jobCallback called
int(9002)
jobCallback called
int(9002)
jobCallback called
int(0)
jobCallback called
int(0)
jobCallback called
int(0)
calling addTask()
job length: 10000
PHP Fatal error:  Allowed memory size of 134217728 bytes exhausted (tried to allocate 4049071631676487512 bytes) in /home/myapp/gearman_client.php on line 42
255
```

In doesn't matter if the workers are running on PHP5 or PHP7, the client is the problem.

**Comment by vojta** [comment]
I'm sorry, but this is still repeatable with the latest version :(

**Comment by wcgallego** [comment]
Hey @vojta,

Made another fix which should clean up the GearmanClient object upon destruction. Give that one a shot too? I'll hold off on closing

**Comment by vojta** [comment]
@wcgallego GearmanClient is NEVER destructed in the testcase above.

**Comment by wcgallego** [comment]
No, but it's destructed upon leaving scope (like when the script ends).

Are you getting excessive memory leaks with this as well? I'm not seeing any memory leaks in valgrind running the above

**Comment by wcgallego** [comment]
Update: ok! I got it reproduced now. Sorry for back and forth. Digging in!

**Comment by vojta** [comment]
When I execute the gearman_client.php, it always reports "Allowed memory size... exhausted".

Are you able to reproduce the same on your system?

```
$ php70 -d memory_limit=1G ./gearman_client.php; echo $?
client created
ping called with result: true
runJobs called
bool(true)
bool(true)
bool(true)
calling addTask()
job length: 10000
OK
runTasks call
jobCallback called
int(10002)
jobCallback called
int(10002)
jobCallback called
int(0)
jobCallback called
int(0)
jobCallback called
int(0)
calling addTask()
job length: 10000
PHP Fatal error:  Allowed memory size of 1073741824 bytes exhausted (tried to allocate 3832905476781323864 bytes) in /home/myapp/gearman_client.php on line 42
255
```

I'm sending strings of about 9.7KB to gearmand and back, but the script consumes more that 1GB and it's not enough. This seems bad to me. But we may have slightly different installs...

**Comment by vojta** [comment]
BTW, I'm using gearmand 1.1.8-2.el6 (Centos 6, 2.6.32-504.30.3.el6.x86_64).

**Comment by wcgallego** [comment]
I did replicate! Sorry for confusion. My guess is I'm doing a poor job of cleaning up a string and it's sending back garbage. So PHP is trying to allocate this huge string without bounds and bailing.

**Comment by wcgallego** [comment]
Hey @vojta,

I got a helping hand from a friend who clean up a good portion of the task code to eliminate some complicated stuff lingering. It's a little easier to sort out now and reduces memory leaks. Should be pushed to trunk now.

There's still one very small leak (96 bytes!) I need to squash, but that should be much easier to figure out and doesn't grow with the number of times things like runJobs() is executed.

I'm leaving this issue open until I can get that last memory leak, but this should get you a little farther if you're waiting on a patch.

**Comment by vojta** [comment]
Indeed, this memory leak is fixed! But, sadly, I get segfaults in other usecases that used to work before :(
I think this extension needs much better automatic test suite, I should add tests instead of reporting these issues separately.

**Comment by rosmo** [comment]
Probably related to the same issue since this referenced both in #5 and #7:

```
Program received signal SIGSEGV, Segmentation fault.
0x0000000000723b75 in _zval_get_string_func ()
(gdb) bt
#0  0x0000000000723b75 in _zval_get_string_func ()
#1  0x0000000000732c9e in zend_is_callable_ex ()
#2  0x000000000071cc90 in zend_call_function ()
#3  0x000000000071d539 in call_user_function_ex ()
#4  0x00007ffff023157a in ?? () from /usr/lib/php/20151012/gearman.so
#5  0x00007ffff000d7b5 in ?? () from /usr/lib/x86_64-linux-gnu/libgearman.so.7
#6  0x00007ffff00061b2 in gearman_client_run_tasks () from /usr/lib/x86_64-linux-gnu/libgearman.so.7
```

The segfault seems to happen only if we do setCompleteCallback() before adding the tasks, not if we do it afterwards (but if it's set afterwards, it's not triggered).

**Comment by acasademont** [comment]
@rosmo i think we're hitting also this new segfaults, master at 9c16acda33f16e3d9ee1dce33431cd0c92566300 didn't have them.

I'm sorry @wcgallego to not being able to provide more meaningfult reports, my experience withe core dumps and gdb is close to 0

**Comment by wcgallego** [comment]
@rosmo, @acasademont can you paste how you're declaring the setCompleteCallback? Are you able to reproduce with just the following?
- declaration of the GearmanClient object
- adding a server with said object
- adding setCompleteCallback
- calling doBackground

Are you passing anything in particular to those functions? What other gearman specific functions are you relying on here?

if it's hitting call_user_function_ex and failing, it's gotta be an empty string that's trying to be allocated and failing. Here's where it's being executed in the gearman extension code:
https://github.com/wcgallego/pecl-gearman/blob/master/php_gearman.c#L2528

And that's part of the client_obj passed in places like this:
https://github.com/wcgallego/pecl-gearman/blob/master/php_gearman.c#L2580

if you can paste any kind of replicable demonstration of the error, that would go a long way! Would love to knock this out for all of you

**Comment by rosmo** [comment]
Hey @wcgallego, can we have a quick discussion on some more realtime media (irc/skype/whatever)? It might be more productive :)

**Comment by wcgallego** [comment]
Sounds like a solid idea! If you have an irc channel in mind, I can hop in now. Am around for the next 2 hours. Usually will_gallego on most servers

Otherwise, online during work week 14:00-22:00 UTC

**Comment by wcgallego** [comment]
FYI - I hopped into #gearman on Freenode. That might be a decent place to look for me!

Can't promise I'll be around at all times, but ping me there anyways

**Comment by wcgallego** [comment]
Ok, so the original intent of this issue was memory leak, which is fixed. Other issues for the package have been opened to address various other problems. I"m going to close this for now and we can either open up a new issue as they arise or if we think the memory issue addressed in the first post is not resolved then reopen this one.

Thanks again to everyone for patience and for contributions.

**Comment by vojta** [comment]
The memory leak is back in the latest code. Today I installed fresh extension in production and the memory usage is extreme again:

Fatal Error: Allowed memory size of 2147483648 bytes exhausted (tried to allocate 319488 bytes)

**Comment by rosmo** [comment]
@vojta, Allocation of 300 kB seems excessively large.. I tried the gearman_worker and gearman_client that you provided and I can't reproduce the error.

**Comment by vojta** [comment]
Maybe it's another memory leak. I know for sure that when I start the client using PHP 5.5, the script never goes over 100M of memory. PHP7 run used 2GB of memory.

**Comment by rosmo** [comment]
@vojta, Do you mean it's reproducible using the scripts you provided earlier? Or some other script? Do you also have some other extensions loaded?

**Comment by vojta** [comment]
@rosmo sorry, I mean our production code, not the testcase which I provided earlier. It's not easy to isolate that, sorry :(

**Comment by rosmo** [comment]
@vojta, Do you have any clue what kind of operation would be allocating 300 kB of memory? Are the workloads you are sending to Gearman so large? Are you creating multiple GearmanClients?

**Comment by vojta** [comment]
@rosmo Yes, we are sending bigger workloads (JPEG images up to 5MB). We are sending them in both directions (the result is an image again). But it has never been problem when running on PHP5.

**Comment by wcgallego** [comment]
hey @vojta, just to confirm - is the issue you're seeing still manifesting from the code snipper in the post post? Also, are you seeing this on the client end or the worker send? @rosmo makes a good point - it's possible it's the gearman extension but it could be another as well.

We're doing tens of millions of jobs a day at my work but we're not seeing the same memory leak. Are you running tasks still or jobs with your Gearman infrastructure? We don't use tasks so that might be a reason why we're not seeing the same memory leak.

**Comment by vojta** [comment]
We're using Tasks, GearmanClient->addTask().
Maybe you're right and it's a problem somewhere else when using PHP7. I have to isolate that case.

**Comment by rosmo** [comment]
@vojta, We are using tasks, but not seeing any big memory leaks anywhere.

**Comment by wcgallego** [comment]
@vojta, if you have a wide array of tasks, can you try running a handful of each and see how memory jumps with each type? That might be easier said than done, but would go a ways in seeing which code is hit

**Comment by wcgallego** [comment]
Hey @vojta,

Fantastic news - had some major help from a coworker who figured out the mem leak! tl;dr circular references are rough.

I've confirmed tests pass and not seeing the mem leak anymore in the code provided. If you can test the issue_19 branch (I'll be merging it into master soon too) I think this should solve your problems!

Major props to @keyurdg for his help here.

**Comment by vojta** [comment]
@wcgallego Thank you, I'll give it a try next week!

**Comment by vojta** [comment]
So, I'm running it in production: it's better, but some memory leak still exists. I'm getting OOM errors with 2GB limit. No problems with PHP 5.5 version with the same memory limit.

Good job, but still not perfect :)

**Comment by wcgallego** [comment]
And Gearman is still the only extension involved here?

I haven't come across any other memory leaks with this yet. Are you able to run an empty task (nothing in the actual work part of it) to eliminate surrounding code?

**Comment by keyurdg** [comment]
@vojta I think this time the fix is in for good :) Check out commit f8bcc7e6df982a97df9bbdcc5e71cfe9eb62223c

**Comment by keyurdg** [comment]
It was very subtle, took a while to figure out what was wrong. The traditional memory leak detection tool valgrind was showing everything as OK.

**Comment by vojta** [comment]
@keyurdg Thank you! It works perfectly so far. 3 days in production without any OOM error.
